datafusion/functions-nested/src/extract.rs

```diff
@@ -993,3 +993,84 @@ where
     let data = mutable.freeze();
     Ok(arrow::array::make_array(data))
 }
+
+#[cfg(test)]
+mod tests {
+    use super::array_element_udf;
+    use arrow_schema::{DataType, Field};
+    use datafusion_common::{Column, DFSchema, ScalarValue};
+    use datafusion_expr::expr::ScalarFunction;
+    use datafusion_expr::{cast, Expr, ExprSchemable};
+    use std::collections::HashMap;
+
+    #[test]
+    fn test_array_element_return_type() {
+        let complex_type = DataType::FixedSizeList(
```

Comment:
When I change this complex type to `DataType::List` the test passes 🤔 

```rust
        let complex_type = DataType::List(
            Field::new("some_arbitrary_test_field", DataType::Int32, false).into(),
        );
```

It also passes when complex_type is a `Struct` 

```rust
        let complex_type = DataType::Struct(Fields::from(vec![
            Arc::new(Field::new("some_arbitrary_test_field", DataType::Int32, false)),
        ]));
```

It seems like there is something about FixedSizeList that is causing issues to me

Reply:
Weird, when I remove this line in expr schema the test passes (with FixedSizedList):
```diff
diff --git a/datafusion/expr/src/expr_schema.rs b/datafusion/expr/src/expr_schema.rs
index 3317deafb..50aeb222f 100644
--- a/datafusion/expr/src/expr_schema.rs
+++ b/datafusion/expr/src/expr_schema.rs
@@ -152,6 +152,7 @@ impl ExprSchemable for Expr {
                     .map(|e| e.get_type(schema))
                     .collect::<Result<Vec<_>>>()?;

+
                 // Verify that function is invoked with correct number and type of arguments as defined in `TypeSignature`
                 let new_data_types = data_types_with_scalar_udf(&arg_data_types, func)
                     .map_err(|err| {
@@ -168,7 +169,7 @@ impl ExprSchemable for Expr {

                 // Perform additional function arguments validation (due to limited
                 // expressiveness of `TypeSignature`), then infer return type
-                Ok(func.return_type_from_exprs(args, schema, &new_data_types)?)
+                Ok(func.return_type_from_exprs(args, schema, &arg_data_types)?)
             }
             Expr::WindowFunction(window_function) => self
                 .data_type_and_nullable_with_window_function(schema, window_function)
```

Which basically says pass the input data types directly to the function call rather than calling `data_types_with_scalar_udf` first (which claims to type coercion)

https://github.com/apache/datafusion/blob/68ead283bd47de75d8aa79f6df5fe78a957df50c/datafusion/expr/src/expr_schema.rs#L171


🤔  this looks like it was added in Sep via https://github.com/apache/datafusion/commit/1b3608da7ca59d8d987804834d004e8b3e349d18 (before that the input types were passed directly) 🤔

Reply:
It doesn't seem right to me that `ExprSchema` is coercing the arguments (implicitly) to me 🤔

Reply:
> It seems like there is something about FixedSizeList that is causing issues to me

correct, https://github.com/apache/datafusion/pull/13756#issuecomment-2542421582

Reply:
> Weird, when I remove this line in expr schema the test passes (with FixedSizedList):

i did the same, basically removing this block
https://github.com/apache/datafusion/blob/b30c2007c2209bca90f2c285a775f09d9a56be03/datafusion/expr/src/expr_schema.rs#L155-L167

it's enough to fix the unit test in this PR
but other things start to fail



> It doesn't seem right to me that `ExprSchema` is coercing the arguments (implicitly) to me 🤔

agreed

Reply:
> It doesn't seem right to me that ExprSchema is coercing the arguments (implicitly) to me 🤔

We need to get the `return_type` of the function here and the arguments of `return_type` is the "coerced data type" therefore I think `new_data_types` is the right choice to me.

Reply:
in a logical plan, the function arguments should already be of the right coerced type, we should just use them.
they may be not required type on the early plan building phase (when plan is still syntactical, not semantical), which unfortuantely uses the same `LogicalPlan` and `Expr` types. https://github.com/apache/datafusion/issues/12604 would address that.

Reply:
> why applying the coercion rules changes the input type in the first place (which seems suspicious)

Fixed size list coerced to List

```rust
    ArrayFunctionSignature::Array => {
        if current_types.len() != 1 {
            return Ok(vec![vec![]]);
        }

        array(&current_types[0])
            .map_or_else(|| vec![vec![]], |array_type| vec![vec![array_type]])
    }

    fn array(array_type: &DataType) -> Option<DataType> {
        match array_type {
            DataType::List(_)
            | DataType::LargeList(_)
            | DataType::FixedSizeList(_, _) => {
                let array_type = coerced_fixed_size_list_to_list(array_type);
                Some(array_type)
            }
            _ => None,
        }
    }
```

Reply:
> This is perhaps the key distinction -- should get_expr_type be applying coercion rules? Or should we insist that the caller applies coercion rules first before calling ExprSchema?

The main issue is that type coercion for function inputs happens after the function is defined. If we had a way to handle coercion before the function is defined, it would significantly simplify the function signature.

Reply:
How about we compute the `return_type` when the function is created, and `get_type` read the value.


```rust
/// ScalarFunction expression invokes a built-in scalar function
#[derive(Clone, PartialEq, Eq, PartialOrd, Hash, Debug)]
pub struct ScalarFunction {
    /// The function
    pub func: Arc<crate::ScalarUDF>,
    /// List of expressions to feed to the functions as arguments
    pub args: Vec<Expr>,

    // Compute the return type when `new_udf()` is called
    // coerced types are computed and become the input of `ScalarUDFImpl::return_type`
    pub return_type: DataType,
}
```

Reply:
> the function arguments should already be of the right coerced type
I don't think this is true because we allow any data types as function's input, and we d the type check or coercion for them.

Reply:
> How about we compute the `return_type` when the function is created, and `get_type` read the value.

I like the idea _in principle_.

It should be combined with a new ScalarUDFImpl sub-trait that doesn't have return type-related methods at all, since they are not to be used once the plan is constructed.


> The reason is because we can't guarantee the input is already coerced.

in a _logical plan_ we can.

My understanding is that coercing analyzer also calls the `get_type` functions.
It can be solved by changing how the coercing analyzer tracks its internal state.

But the real problem is that same types, the `LogicalPlan` & `Expr`, have two meanings: syntactic and semantic. So in the code we go back and forth about what should and what cannot be guaranteed for an `Expr` or `LogicalPlan` instance.

Reply:
> the LogicalPlan & Expr, have two meanings: syntactic and semantic.

Is there example about the difference of this two, especially for function. For Expr::ScalarFunction, it has no difference in LogicalPlan, we don't do anything special, but I think this is what you don't expect. What should we have in LogicalPlan, `Expr::ScalarFunction` but with coerced input?

Reply:
> since they are not to be used once the plan is constructed.

Why `get_type` is not supposed to be available after plan is constructed from Expr.

Reply:
> Is there example about the difference of this two, especially for function.

the difference is more apparent for duplicate syntax (such is IS NULL vs IS UNKNOWN), syntax sugar (order by 1, order by all, select *)
for function call the difference is about function being resolved (typed and inputs coerced) or not.


> > since they are not to be used once the plan is constructed.
> 
> Why `get_type` is not supposed to be available after plan is constructed from Expr.

for a fully resolved logical plan it's fair question to ask what is the type of an expression (and this may or may not be O(1) available answer)

however, there is no point to ask a UDF what is its type, since we already asked it 

think of this as engine and UDF being implemented by independent parties, with UDF being a contract layer.
you go over a contract layer when you have to (analysis time), but going over contract layer multiple times with the same question should be avoided.